File: src/main/java/com/spacemike/extradevices/template/AdvDroneTemplate.java
package com.spacemike.extradevices.template;

import li.cil.oc.api.*;
import li.cil.oc.api.network.EnvironmentHost;
import li.cil.oc.common.Slot;
import li.cil.oc.common.Tier;
import li.cil.oc.util.ItemUtils;
import net.minecraft.inventory.IInventory;
import net.minecraft.item.ItemStack;
import java.util.ArrayList;

import org.apache.commons.lang3.tuple.Pair;

//import scala.collection.convert.WrapAsJava._

public class AdvDroneTemplate {
	
	/*public final Class<? extends EnvironmentHost> hostClass() {
		return li.cil.oc.api.internal.Drone.class;
	}*/
	public static boolean selectTier3(ItemStack stack) {
		System.out.println("itemtest");
		System.out.println(Items.get("dronecase3"));
		System.out.println(Items.get(stack));
		return (Items.get(stack) == Items.get("dronecase3"));
	}

	public static void register() {
		ArrayList<Pair<String, Integer>> containerSlots = new ArrayList<Pair<String, Integer>>();
		containerSlots.add(Pair.of(Slot.Card(), Tier.Three()));
		containerSlots.add(Pair.of(Slot.Card(), Tier.Two()));
		containerSlots.add(Pair.of(Slot.Card(), Tier.Two()));
		containerSlots.add(Pair.of(Slot.CPU(), Tier.Two()));
		containerSlots.add(Pair.of(Slot.Memory(), Tier.Two()));
		containerSlots.add(Pair.of(Slot.Memory(), Tier.Two()));
		containerSlots.add(Pair.of(Slot.EEPROM(), Tier.Any()));
		
		Class<? extends EnvironmentHost> hostClass = li.cil.oc.api.internal.Drone.class;
		
		// Tier 3
	    IMC.registerAssemblerTemplate(
	      "Drone (Tier 3)",
	      "com.spacemike.extradevices.template.AdvDroneTemplate.selectTier3",
	      "li.cil.oc.common.template.DroneTemplate.validate",
	      "li.cil.oc.common.template.DroneTemplate.assemble",
	      hostClass,
	      null,
	      new int[] {
	    	Tier.Three(),
	    	Tier.Three(),
	    	Tier.Two(),
	    	Tier.Two(),
	    	Tier.One()
	      },
	      //
	      containerSlots
	      );
		
		/*NBTTagCompound nbt = new NBTTagCompound();
		nbt.setString("select", "com.spacemike.extradevices.template.AdvDroneTemplate.selectTier3");
		nbt.setString("validate", "li.cil.oc.common.template.DroneTemplate.validate");
		nbt.setString("assemble", "li.cil.oc.common.template.DroneTemplate.assemble");
		nbt.setString("hostClass", "com.spacemike.extradevices.template.AdvDroneTemplate.hostClass");
		NBTTagList upgslots = new NBTTagList();
		NBTTagCompound upgslot1 = new NBTTagCompound();
		upgslot1.setString("type", value);
		upgslots.appendTag(nbt);*/
	}

	public int maxComplexity(IInventory inventory) {
		if (caseTier(inventory) == 2) return 16;
		return 5;
	}

	public int caseTier(IInventory inventory) {
		return ItemUtils.caseTier(inventory.getStackInSlot(0));
	}
}

File: src/main/scala/li/cil/oc/common/template/AdvDroneTemplate.scala
package li.cil.oc.common.template

object AdvDroneTemplate extends DroneTemplate {
  //override protected val suggestedComponents = Array(
  //  "BIOS" -> hasComponent(Constants.ItemName.EEPROM) _)

  //override protected def hostClass = classOf[internal.Drone]

  //def selectTier3(stack: ItemStack) = api.Items.get(stack) == api.Items.get(Constants.ItemName.DroneCaseTier3)
  public bool selectTier3(stack: ItemStack){
    return api.Items.get(stack) == api.Items.get("dronecase3");
  }

  /*def validate(inventory: IInventory): Array[AnyRef] = validateComputer(inventory)

  def assemble(inventory: IInventory) = {
    val items = (0 until inventory.getSizeInventory).map(inventory.getStackInSlot)
    val data = new DroneData()
    data.tier = caseTier(inventory)
    data.name = RobotData.randomName
    data.components = items.drop(1).filter(!_.isEmpty).toArray
    data.storedEnergy = Settings.get.bufferDrone.toInt
    val stack = api.Items.get(Constants.ItemName.Drone).createItemStack(1)
    data.save(stack)
    val energy = Settings.get.droneBaseCost + complexity(inventory) * Settings.get.droneComplexityCost

    Array(stack, Double.box(energy))
  }

  def selectDisassembler(stack: ItemStack) = api.Items.get(stack) == api.Items.get(Constants.ItemName.Drone)

  def disassemble(stack: ItemStack, ingredients: Array[ItemStack]) = {
    val info = new MicrocontrollerData(stack)
    val itemName = Constants.ItemName.DroneCase(info.tier)

    Array(api.Items.get(itemName).createItemStack(1)) ++ info.components
  }*/

  def register() {
    // Tier 1
    api.IMC.registerAssemblerTemplate(
      "Drone (Tier 3)",
      "li.cil.oc.common.template.DroneTemplate.selectTier3",
      "li.cil.oc.common.template.DroneTemplate.validate",
      "li.cil.oc.common.template.DroneTemplate.assemble",
      hostClass,
      null,
      Array(
        Tier.Three,
        Tier.Three,
        Tier.Two,
        Tier.Two,
        Tier.One,
      ),
      asJavaIterable(Iterable(
        (Slot.Card, Tier.Three),
        (Slot.Card, Tier.Two),
        (Slot.Card, Tier.Two),
        null,
        (Slot.CPU, Tier.Two),
        (Slot.Memory, Tier.Two),
        null,
        (Slot.EEPROM, Tier.Any)
      ).map(toPair)))

    // Disassembler
    /*api.IMC.registerDisassemblerTemplate(
      "Drone",
      "li.cil.oc.common.template.DroneTemplate.selectDisassembler",
      "li.cil.oc.common.template.DroneTemplate.disassemble")*/
  }

  override protected def maxComplexity(inventory: IInventory) =
    if (caseTier(inventory) == Tier.Three) 16

  override protected def caseTier(inventory: IInventory) = ItemUtils.caseTier(inventory.getStackInSlot(0))
}